ENH: Improved error message and raise new error for small-string NaN edge case in HDFStore.append

JakeTT404 opened the pull request:
Updated **pytables.py** to improve error messages caused by datatype mismatches in **HDFStore.append**. Also added **ValueError** for when the NaN representation cannot fit into the column. Modified tests concerning the improved error message and added new test for when column is type string with length <3 and as such **nan_rep** 'nan' is too big.

- [ ] closes #16300
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests)
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).

JakeTT404 committed:
Add clearer error messages for datatype mismatch in HDFStore.append. Raise ValueError when nan_rep too large for pytable column. Add and modify applicable test code.

JakeTT404 committed:
Fix missed tests and correct mistake in error message.

mroeschke commented on pandas/tests/io/pytables/test_append.py:
```diff
@@ -421,6 +421,14 @@ def check_col(key, name, size):
             store.append("df_new", df_new)
 
+        # bigger NaN representation on next append
+        df_new = DataFrame([[124, "a"], [346, "b"]])
```
Can you make this case a separate test function?

mroeschke commented on pandas/io/pytables.py:
```diff
@@ -3524,6 +3524,14 @@ def validate(self, other) -> None:
                     # Value of type "Optional[Any]" is not indexable  [index]
                     oax = ov[i]  # type: ignore[index]
                     if sax != oax:
+                        ## Raise clearer error if mismatching type on values_axes
```
```suggestion
```

IMO this comment doesn't add much

mroeschke commented on pandas/io/pytables.py:
```diff
@@ -3524,6 +3524,14 @@ def validate(self, other) -> None:
+                                f"because its data contents are not [{sax.kind}] "
+                                f"but [{oax.kind}] object dtype"
+                            )
+                        # Fallback if other source of difference
```
```suggestion
```

mroeschke commented on pandas/tests/io/pytables/test_append.py:
```diff
@@ -823,14 +831,11 @@ def test_append_raise(setup_path):
+            "but [datetime64[s]] object dtype"
         )
-        with pytest.raises(ValueError, match=msg):
+        with pytest.raises(TypeError, match=msg):
```
I don't think the exception type should change. That would be an API change

JakeTT404 committed:
Remove excess comments. Reverse error type change to avoid api changes. Move nan_rep tests into separate function.

JakeTT404 committed:
Merge remote-tracking branch 'upstream/main' into Wrong-error-message-in-HDFStore.append

JakeTT404 commented:
I've removed the comments mentioned and reverted the error type. I've replaced the test with a function of 3 tests testing the following cases:

- Error on string column too small
- Error on modified nan_rep too big
- Success on small modified nan_rep in small string column

There may be some odd git stuff during me updating the main branch due to my unfamiliarity with the GitHub contribution workflow and git as a whole.

JakeTT404 committed:
Merge remote-tracking branch 'upstream/main' into Wrong-error-message-in-HDFStore.append

mroeschke commented:
Thanks @JakeTT404

rhshadrach commented:
This impacted #60795 - https://github.com/pandas-dev/pandas/pull/60795#discussion_r1943799181

Is this okay to backport?

mroeschke commented:
Yup sure OK with me

rhshadrach commented:
Backport PR: #60907
